import grpc
from concurrent import futures
import time

# å¯¼å…¥ç”Ÿæˆçš„ä»£ç 
import vision_pb2
import vision_pb2_grpc

# å¯¼å…¥ä¸šåŠ¡é€»è¾‘
from core.embedding_service import embedding_service
from core.ocr_service import ocr_service
from utils.image_loader import load_image_from_url
from core.caption_service import caption_service



class VisionServer(vision_pb2_grpc.VisionServiceServicer):

    def EmbedText(self, request, context):
        try:
            print(f"ğŸ“ Request EmbedText: {request.text}")
            vector = embedding_service.embed_text(request.text)
            return vision_pb2.EmbeddingResponse(vector=vector, dim=len(vector))
        except Exception as e:
            print(f"Error: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(str(e))
            return vision_pb2.EmbeddingResponse()

    def EmbedImage(self, request, context):
        try:
            print(f"ğŸ–¼ï¸ Request EmbedImage: {request.url}")
            # 1. ä¸‹è½½å›¾ç‰‡
            image = load_image_from_url(request.url)
            # 2. è®¡ç®—å‘é‡
            vector = embedding_service.embed_image(image)
            return vision_pb2.EmbeddingResponse(vector=vector, dim=len(vector))
        except Exception as e:
            print(f"Error: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(str(e))
            return vision_pb2.EmbeddingResponse()

    def ExtractText(self, request, context):
        try:
            print(f"ğŸ” Request OCR: {request.url}")
            image = load_image_from_url(request.url)
            full_text = caption_service.extract_text_ocr(image)
            lines = full_text.split('\n')
            lines = [line.strip() for line in lines if line.strip()]
            return vision_pb2.OcrResponse(full_text=full_text, lines=lines)
        except Exception as e:
            print(f"Error: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(str(e))
            return vision_pb2.OcrResponse()

    def GenerateCaption(self, request, context):
        """
        æµå¼ç”Ÿæˆæ¥å£
        request.image_url: å›¾ç‰‡é“¾æ¥
        request.prompt: (å¯é€‰) æ¯”å¦‚ "å†™ä¸€ä¸ªå°çº¢ä¹¦æ–‡æ¡ˆ"
        """
        try:
            print(f"âœ¨ Request Gen: {request.image_url}")

            # å¦‚æœ Java ç«¯æ²¡ä¼  promptï¼Œç»™ä¸ªé»˜è®¤å€¼
            prompt = request.prompt if request.prompt else "è¯·è¯¦ç»†æè¿°è¿™å¼ å›¾ç‰‡"

            # è°ƒç”¨æµå¼ç”Ÿæˆ
            for chunk in caption_service.stream_generate(request.image_url, prompt):
                # å®æ—¶è¿”å›ç»™ Java
                yield vision_pb2.StringResponse(content=chunk)

        except Exception as e:
            print(f"Error: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(str(e))
            # æµå¼æ¥å£å‡ºé”™ä¹Ÿå¾— yield ä¸€ä¸ªç©ºæˆ–è€…é”™è¯¯ä¿¡æ¯ï¼Œé˜²æ­¢å®¢æˆ·ç«¯å¡æ­»
            yield vision_pb2.StringResponse(content=f"[Error: {str(e)}]")


def serve():
    # åˆ›å»º gRPC æœåŠ¡å™¨ï¼Œä½¿ç”¨çº¿ç¨‹æ±  (æœ€å¤§10å¹¶å‘)
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))

    # æ³¨å†ŒæœåŠ¡
    vision_pb2_grpc.add_VisionServiceServicer_to_server(VisionServer(), server)

    # ç›‘å¬ç«¯å£
    port = '[::]:50051'
    server.add_insecure_port(port)
    print(f"âœ… gRPC Server started on {port}")
    print("   - Chinese-CLIP (768 dim)")
    print("   - PaddleOCR (v4)")

    server.start()

    # ä¿æŒè¿è¡Œ
    try:
        while True:
            time.sleep(86400)
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == '__main__':
    serve()